src/python/dxpy/sugar/xfer.py

```diff
@@ -0,0 +1,806 @@
+    project=None, block=True
+):
+    """
+    Download and decompress a gzipped file.
+
+    Args:
+        remote_file (dxpy.DXFile or dxlink): DNAnexus link or file-id of file to
+            download.
+        input_filename (str): Name to use for the input filename, if different than
+            the name of the input_file. If not provided, platform filename is used.
+        local_filename: Local filename/dirname. If not None, this file/directory must
+            exist after unpacking the archive or an error is raised.
+        output_dir (str): Download file to a specific directory (default is the current
+            directory).
+        project: The project that contains the file, if it is not the currently
+            selected project and is not specified in the remote file object/link.
+        block: Wait for the download to complete before returning.
+
+    Returns:
+        The path of the decompressed file. Also returns the running download process if
+        `block is False`.
+    """
+    remote_file = _as_dxfile(remote_file, project)
+
+    if input_filename is None:
+        input_filename = remote_file.describe()["name"]
+
+    if local_filename is None:
+        local_filename = input_filename[:-3]
+    if output_dir:
+        local_filename = os.path.join(output_dir, local_filename)
+
+    cmds = [
+        ["dx", "download", remote_file.get_id(), "-o", "-"],
+        ["gunzip", ">", local_filename]
+    ]
+
+    info(
+        "Downloading file {} to {} using command {}",
+        remote_file.get_id(), local_filename, cmds
+    )
+
+    result = chain(cmds, block=block)
+
+    if block:
+        return local_filename
+    else:
+        return local_filename, result[-1]
+
+
+def _as_dxfile(fileobj, project=None):
+    if dxpy.is_dxlink(fileobj):
+        project_id = None
+        if (
+            isinstance(fileobj["$dnanexus_link"], dict) and
+            "project" in fileobj["$dnanexus_link"]
+        ):
+            project_id = fileobj["$dnanexus_link"]["project"]
+        elif project:
+            project_id = project.get_id()
+        return dxpy.DXFile(fileobj, project=project_id)
+    if not isinstance(fileobj, dxpy.DXFile):
+        raise ValueError("Not a link or DXFile object: {}".format(fileobj))
+    return fileobj
+
+
+class DataTransferExecutor(concurrent.futures.ThreadPoolExecutor):
+    """
+    Abstract subclass of :class:`concurrent.futures.ThreadPoolExecutor` that manages
+    processing files in a multithreaded manner.
+
+    Args:
+        io_function: Function to call to perform the data transefer.
+        max_num_parallel_transfers: Maximum number of parallel threads.
+        default_kwargs: Keyword arguments to pass to every enqueue call.
+    """
+    def __init__(self, io_function, max_parallel=None, **default_kwargs):
+        super(DataTransferExecutor, self).__init__(
+            max_workers=num_cores(max_parallel)
```

Comment:
A question more for discussion:

Often times clients just make API calls and use bandwidth. A lot of the work performed is Network I/O bound. In practice, I see people use a heuristic of 2 * `num_cores` when dealing with network I/O bound task.

For this scenario, uploading/downloading data, what are good heuristics for `ThreadPoolExecutor` worker count